File: src/chapters/linear-systems/content/code/back-substitution.cpp
#include <vector>
#include <iostream>
using namespace std;
using Vec = vector<double>;
using Mat = vector<Vec>;

// Solve an upper-triangular system U x = b.
Vec back_substitution(const Mat& U, const Vec& b) {
    int n = b.size();
    Vec x(n);
    for (int i = n - 1; i >= 0; --i) {
        double s = b[i];
        for (int j = i + 1; j < n; ++j) s -= U[i][j] * x[j];
        x[i] = s / U[i][i];
    }
    return x;
}

int main() {
    Mat U = {{2, 1, -1}, {0, 1, 2}, {0, 0, 3}};
    Vec b = {1, 8, 9};
    for (double v : back_substitution(U, b)) cout << v << " ";
    cout << "\n";
}

File: src/chapters/linear-systems/content/code/back-substitution.rs
// Solve an upper-triangular system U x = b.
fn back_substitution(u: &[Vec<f64>], b: &[f64]) -> Vec<f64> {
    let n = b.len();
    let mut x = vec![0.0; n];
    for i in (0..n).rev() {
        let mut s = b[i];
        for j in i + 1..n { s -= u[i][j] * x[j]; }
        x[i] = s / u[i][i];
    }
    x
}
fn main() {
    let u = vec![vec![2.0, 1.0, -1.0], vec![0.0, 1.0, 2.0], vec![0.0, 0.0, 3.0]];
    let b = vec![1.0, 8.0, 9.0];
    println!("{:?}", back_substitution(&u, &b));   // [1, 2, 3]
}
